docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,435 @@
+            self.trace_path.into_iter().fold(String::new(), |acc, val| {
+                let next = Self::get_denom_prefix(&val.port_id, &val.channel_id);
+                format!("{}{}", acc, next)
+            })
+        }
+        /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+        /// 'ibc/trace_path/base_denom'. If the trace is empty, it will return the base denomination.
+        fn ibc_denom(&self) -> String {
+            format!("ibc/{}", self.get_full_denom_path())
+        }
+        /// Returns the prefix for this trace
+        fn get_prefix(&self) -> String {
+            DenomTrace::get_denom_prefix(self.trace_path[0].port_id, self.trace_path[0].channel_id)
+        }
+
+        /// Returns the prefix for this trace
+        pub fn has_prefix(denom: String, prefix: String) -> bool {
+            denom.starts_with(prefix)
+        }
+
+        /// get_denom_prefix returns the receiving denomination prefix
+        pub fn get_denom_prefix(port_id: &PortId, channel_id: &ChannelId) -> String {
+            format!("{}/{}/", port_id, channel_id)
+        }
+
+
+        /// get_prefixed_denom returns the denomination with the port_id and channel_id prefixed
+        pub fn get_prefixed_denom(port_id: &PortId, channel_id: &ChannelId, base_denom: String) -> String {
+            format!("{}/{}/{}", port_id, channel_id, base_denom)
+        }
+        
+        pub fn parse_denom_trace(denom: &str) -> Result<Self, ICS20Error> {
+            
+        }
+    }
+
+    impl FromStr for DenomTrace {
+        type Error = ICS20Error
+        fn from_str(val: &str) -> Result<Self, Self::Error> {
+            DenomTrace::parse_denom_trace(val)
+        }
+    }
+
+    /// sender_chain_is_source returns false if the denomination originally came
+    /// from the receiving chain and true otherwise.
+    fn sender_chain_is_source(source_port: &PortId, source_channel: &ChannelId, denom: String) -> bool {
+        // This is the prefix that would have been prefixed to the denomination
+        // on sender chain IF and only if the token originally came from the
+        // receiving chain.
+
+        !receiver_chain_is_source(source_port, source_channel, denom)
+    }
+
+    /// receiver_chain_is_source returns true if the denomination originally came
+    /// from the receiving chain and false otherwise.
+    fn receiver_chain_is_source(source_port: &PortId, source_channel: &ChannelId, denom: String) -> bool {
+        // The prefix passed in should contain the SourcePort and SourceChannel.
+        // If  the receiver chain originally sent the token to the sender chain
+        // the denom will have the sender's SourcePort and SourceChannel as the
+        // prefix.
+
+        let voucher_prefix = get_denom_prefix(source_port, source_channel);
+        DenomTrace::has_prefix(denom, voucher_prefix)
+
+    }
+
+    pub trait ICS20Keeper: BankKeeper<Self::AccountId> 
+        + AccountKeeper<Self::AccountId>
```

Comment:
`AccountKeeper` is never defined

Reply:
Up. Let's clarify why is `AccountKeeper` necessary here.